in_podman_metrics: Initial plugin implementation

pabloxxl opened the pull request:
This patch introduces new plugin called in_podman_metrics. It reads list of
containers from podman internal storage, and for each container, reads
information from /sys and /proc containers (so no podman command is ever
executed). After that, it exposes this data in cmetric format, so exporters
like prometheus_exporter can grab it.

This patch introduces initial version of in_podman_metrics plugin - for now it
exposes 10 types of metrics: 4 related to container memory usage, 2 related to
container cpu usage and 4 related to container network usage.

Signed-off-by: Paweł Cendrzak <pabloxxl@fastmail.fm>

<!-- Provide summary of changes -->
This PR adds initial implementation of in_podman_metrics input plugin. This plugin grabs podman container list from podman configuration json file, and for each container grabs information about cpu, memory and network usage from /sys and /proc filesystems. It supports both versions v1 and v2 of cgroups, and can detect containers in nested directories for both versions.
This plugin does not use any podman commands or API calls.

Each scrape consists of four stages:
1) Read config json file (defaulting /var/lib/containers/storage/overlay-containers/containers.json). Grab name and if for each entry
2) For each container, localize /sys/fs/cgroups directory (directories) that contain its data. Save memory and cpu information. Also check first PID connected with it.
3) For each container, get network usage from /proc/<pid>/dev/net by using PID from step 2
4) For each container, create counter (or gauge) and set its value

The code included in this Pull Request was developed during implementation of Motorola Solutions metrics feature. It mimics part of functionalities delivered by cadvisor (but for podman containers, which are not supported by it). Because of it, metric names mimic cadvisor naming convention.

Valgrind execution log [valgrind.txt](https://github.com/fluent/fluent-bit/files/9126075/valgrind.txt)
Configuration file [conf.txt](https://github.com/fluent/fluent-bit/files/9126080/conf.txt)
Log with debug log level [log.txt](https://github.com/fluent/fluent-bit/files/9126088/log.txt)
[Documentation PR](https://github.com/fluent/fluent-bit-docs/pull/844)

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [N/A] Attached [local packaging test](./packaging/local-build-all.sh) output showing all targets (including any new ones) build.

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->
[Documentation PR](https://github.com/fluent/fluent-bit-docs/pull/844)


**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [N/A] Backport to latest stable release.

<!--  Other release PR (not required but highly recommended for quick turnaround) -->
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

patrick-stephens commented:
@pabloxxl can you rebase, we were having issues with the MacOS unit tests that are hopefully resolved now? Yours timeout here.

pabloxxl commented:
Whoops, Sorry, I thought that "comment and close" meant closing particular comment, not entire pull request :(

patrick-stephens commented on plugins/in_podman_metrics/podman_metrics.h:
```diff
@@ -0,0 +1,88 @@
+
+static struct flb_config_map config_map[] = {
+    {
+     FLB_CONFIG_MAP_TIME, "scrape_interval", "20",
```
Any reason for this default? 

I had a quick look and it appears we have scrape_interval defaults currently of 1,2,5,10 and 30 seconds depending on the plugin used. Personally I would prefer we try to keep to one of the existing ones (e.g. 10s) and reduce this sprawl but it is only a nit pick.

pabloxxl replied:
I'm having trouble to remember why I picked this particular value - I've changed it to 30s

pabloxxl replied:
As for unit tests: I've checked it when i started to write this plugin, but it seems there are no tests (so no framework, no functions etc) for metric related plugins. Do you think it is still worth the effort? I can unit test particular mechanisms without checking counter values from "outside"

pabloxxl replied:
Right, I've added unit tests for my plugin - they generate metrics from various sources (or their absence). Since it is metric plugin and it doesn't publish data like "normal ones", I had to dig into plugin internals in my tests (in order to get harvested metrics).

Attaching log file from unit test execution:
[unit_tests.log](https://github.com/fluent/fluent-bit/files/9541440/unit_tests.log)

patrick-stephens replied:
Thanks @pabloxxl

patrick-stephens commented on CMakeLists.txt:
```diff
@@ -166,6 +166,7 @@ option(FLB_IN_STORAGE_BACKLOG          "Enable storage backlog input plugin"
 option(FLB_IN_EMITTER                  "Enable emitter input plugin"                  Yes)
 option(FLB_IN_NODE_EXPORTER_METRICS    "Enable node exporter metrics input plugin"    Yes)
 option(FLB_IN_WINDOWS_EXPORTER_METRICS "Enable windows exporter metrics input plugin" Yes)
+option(FLB_IN_PODMAN_METRICS           "Enable Podman Metrics input plugin"           Yes)
```
As this plugin is disabled for Windows, we should add that explicitly to the `cmake/windows-setup.cmake` set of defaults. Similar for MacOS - it only functions on Linux looking at the code.

pabloxxl replied:
I've changed windows-setup.cmake one, but I cannot find similar list for macos build - can you point me into a right direction?

patrick-stephens replied:
I think it is this one @pabloxxl but I would ask @cosmo0920 : https://github.com/fluent/fluent-bit/blob/master/cmake/macos-setup.cmake
I don't have a Mac system to test but unit tests do run on it.

cosmo0920 replied:
How about adding on the macos-setup.cmake and adding somewhat needed for this operaion?

```cmake
if(FLB_MACOS_DEFAULTS)
  message(STATUS "Overriding setttings with macos-setup.cmake")
  set(FLB_IN_PODMAN_METRICS No)
endif ()
```

pabloxxl replied:
Should I add more plugins inside this section? I'm sure mine plugin is not the only one that should be disabled on mac

cosmo0920 replied:
Ah, I'd misunderstood for cmake configuration of your plugin.
It is already only for Linux: https://github.com/fluent/fluent-bit/pull/5732/files#diff-1abf564a84becf73b24c833a7300a9fcaf5ff9df6427cb54fd53b297806669a0R173 (Inside Linux only block).
So, we needn't further handling for macOS specific.

edsiper commented on plugins/in_podman_metrics/podman_metrics.h:
```diff
@@ -0,0 +1,98 @@
+     0, FLB_TRUE, offsetof(struct flb_in_metrics, config_podman_config),
+     "Path to podman config file"
+    },
+    {
```
for the config options that sets a path, let's use the same config variable names found in node_exporter_metrics, users are aware of those, e.g:

https://github.com/fluent/fluent-bit/blob/master/plugins/in_node_exporter_metrics/ne.c#L163-L172

edsiper commented on plugins/in_podman_metrics/podman_metrics.c:
```diff
@@ -0,0 +1,496 @@
+    jsmn_parser p;
+    jsmntok_t t[JSON_TOKENS];
+
+    fp = fopen(ctx->config, "rb");
```
please use `int flb_utils_read_file(char *path, char **out_buf, size_t *out_size)` to read a file

edsiper commented on plugins/in_podman_metrics/podman_metrics.c:
```diff
@@ -0,0 +1,496 @@
+
+    for (i=0; i<r; i++) {
+        if (t[i].type == JSMN_STRING) {
+            if ((int)strlen(JSON_FIELD_ID) == t[i].end - t[i].start &&
```
instead of strlen() you can do `sizeof(JSON_FIELD_ID) - 1` so the value get statically set, strlen will run on every loop

edsiper commented on plugins/in_podman_metrics/podman_metrics.c:
```diff
@@ -0,0 +1,496 @@
+                strncmp(buffer + t[i].start, JSON_FIELD_ID, t[i].end - t[i].start) == 0) {
+                memset(id, 0, sizeof(id));
+                strncpy(id, buffer + t[i+1].start, t[i + 1].end - t[i + 1].start);
+                id[strlen(id)] = '\0';
```
if `id` would not have a `\0` in place before, strlen would make it crash. I suggest to calculate the length of what you are copying in strncpy in a variable, e.g:

```c
len = t[i + 1].end - t[i + 1].start;
```

then just do `id[len] = '\0';`

with this approach, the memset() used before is not longer required.

edsiper commented on plugins/in_podman_metrics/podman_metrics.c:
```diff
@@ -0,0 +1,496 @@
+            else if ((int)strlen(JSON_FIELD_NAMES) == t[i].end - t[i].start &&
+                     strncmp(buffer + t[i].start, JSON_FIELD_NAMES, t[i].end - t[i].start) == 0) {
+                array_id = i + 1;
+                if (t[array_id].type == JSMN_ARRAY)
```
open `{` in the same line that `if` exists (not the next line):

ref: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

edsiper commented on plugins/in_podman_metrics/podman_metrics.c:
```diff
@@ -0,0 +1,496 @@
+                    j = array_id + 1;
+                    while (t[j].parent == array_id)
+                    {
+                        memset(name, 0, sizeof(name));
```
remove memset, calculate length and set it manually

edsiper commented on plugins/in_podman_metrics/podman_metrics.c:
```diff
@@ -0,0 +1,496 @@
+static int add_container_to_list(struct flb_in_metrics *ctx, flb_sds_t id, flb_sds_t name)
+{
+    struct container *cnt;
+    cnt = flb_malloc(sizeof(struct container));
```
validate that cnt succeed: 

```c
if (!cnt) {
    flb_errno();
    ...
}
```

edsiper commented on plugins/in_podman_metrics/podman_metrics.c:
```diff
@@ -0,0 +1,496 @@
+    struct container *cnt;
+    struct net_iface *iface;
+    struct sysfs_path *pth;
+    mk_list_foreach_safe(ctx->head, ctx->tmp, &ctx->items) {
```
it's not necessary to have ctx->`struct mk_list tmp` or cnt->`struct mk_list tmp`, just declare those list iterators in the function that you need them, not in the contexts.

edsiper commented on plugins/in_podman_metrics/podman_metrics_data.c:
```diff
@@ -0,0 +1,420 @@
+ * Destroy flb counter instance, and set pointer to it to NULL (so checks later
+ * can determine if it was removed.
+ */
+int destroy_counter(struct flb_in_metrics *ctx, struct cmt_counter **c)
```
I don't think this function is useful, actually having a debug message won't help. I suggest to remove this function.

edsiper commented on plugins/in_podman_metrics/podman_metrics_data.c:
```diff
@@ -0,0 +1,420 @@
+ * Destroy flb gauge instance, and set pointer to it to NULL (so checks later
+ * can determine if it was removed.
+ */
+int destroy_gauge(struct flb_in_metrics *ctx, struct cmt_gauge **g)
```
same as the other, just remove this function

edsiper commented on plugins/in_podman_metrics/podman_metrics_data.c:
```diff
@@ -0,0 +1,420 @@
+            switch (curr_field++)
+            {
+            case DEV_NET_NAME:
+                // Remove ':' from the end of name
```
/* ... */

ref: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#single-line-comment

edsiper commented on plugins/in_podman_metrics/podman_metrics_data.c:
```diff
@@ -0,0 +1,420 @@
+            if (name_starts_with(entry->d_name, SYSFS_CONTAINER_PREFIX) == 0 &&
+                strcmp(entry->d_name, SYSFS_LIBPOD_PARENT) != 0 &&
+                strstr(entry->d_name, SYSFS_CONMON) == 0) {
+                pth = flb_malloc(sizeof(struct sysfs_path));
```
validate that allocation succeeded

edsiper reviewed (changes requested):
thank you for this contribution! overall looks pretty good. I did a quick review and I will ask for some small changes and extra validations. 

Definitely this is something we should have as part of v2.0 release.

pabloxxl commented:
I've noticed that all builds in workflows failed because cmt_time_now() is now cfl_time_now(). I hope it will work now.

pabloxxl commented:
Also, in of TCs on ubuntu machine I've noticed segmentation fault while assigning numer of read bytes... It seems that this variable should be size_t (d'oh) and not int - so I've changed it.

pabloxxl commented:
@edsiper Hey again, just reminding about this PR

patrick-stephens commented:
@pabloxxl I've triggered the packaging tests just to confirm it builds for all targets.

pabloxxl commented:
It seems that a lot of build targets failed :( I have backmerged to master and removed .event_type (since it looks like it is no longer used for input plugins)

patrick-stephens commented:
@pabloxxl could you update your commits to be signed off and use the `in_podman_metrics: ` prefix? I can approve the CI runs but it'll just fail because of that so be good to get it in first.

pabloxxl commented:
> @pabloxxl could you update your commits to be signed off and use the `in_podman_metrics: ` prefix? I can approve the CI runs but it'll just fail because of that so be good to get it in first.

I've changed commit from the top to use prefix and be signed - or should I change every one of my commits?
I can also squash them I guess.

pabloxxl commented:
> > @pabloxxl could you update your commits to be signed off and use the `in_podman_metrics: ` prefix? I can approve the CI runs but it'll just fail because of that so be good to get it in first.
> 
> I've changed commit from the top to use prefix and be signed - or should I change every one of my commits? I can also squash them I guess.

I've squashed all commits into a original one

patrick-stephens commented:
> @edsiper @patrick-stephens it seems that all checks are passed. Is there anything more to be done here?

Just needs @edsiper to confirm he is happy with the changes since review I think.

patrick-stephens commented:
@pabloxxl are you willing to take ownership of maintaining this? We need to specify codeowners as well.

pabloxxl commented:
> @pabloxxl are you willing to take ownership of maintaining this? We need to specify codeowners as well.

Sure! I will be glad if could be more involved.

pabloxxl commented:
> > @edsiper @patrick-stephens it seems that all checks are passed. Is there anything more to be done here?
> 
> Just needs @edsiper to confirm he is happy with the changes since review I think.

Hey @edsiper, can you take a look?
